src/AutoRest.CSharp/Common/Output/Models/Types/HelperTypeProviders/ChangeTrackingListProvider.cs

```diff
@@ -0,0 +1,282 @@
+        private VariableReference _innerList;
+
+        public static ChangeTrackingListProvider Instance => _instance.Value;
+
+        private ChangeTrackingListProvider(string defaultNamespace, SourceInputModel? sourceInputModel)
+            : base(defaultNamespace, sourceInputModel)
+        {
+            _innerlListField = new FieldDeclaration(FieldModifiers.Private, typeof(IList<>), "_innerList");
+            _innerList = new VariableReference(new CSharpType(typeof(IList<>)), _innerlListField.Declaration);
+            _t = new GenericParameterTypeProvider("T", DefaultNamespace, null);
+            _tArray = new GenericParameterTypeProvider("T[]", DefaultNamespace, null);
+            _tParam = new Parameter("item", null, _t.Type, null, ValidationType.None, null);
+            DeclarationModifiers = ClassSignatureModifiers.Internal;
+        }
+
+        protected override string DefaultName => "ChangeTrackingList";
+
+        protected override IEnumerable<Method> BuildConstructors()
+        {
+            yield return new Method(new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, Array.Empty<Parameter>()), Snippets.EmptyStatement);
+            var iListParam = new Parameter("innerList", null, typeof(IList<>), null, ValidationType.None, null);
+            var iListSignature = new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, new Parameter[] { iListParam });
+            var iListVariable = new ParameterReference(iListParam);
+            var iListbody = new MethodBodyStatement[]
+            {
+                new IfStatement(Snippets.NotEqual(iListVariable, Snippets.Null))
+                {
+                    new AssignValueStatement(_innerList, iListVariable)
+                }
+            };
+
+            yield return new Method(iListSignature, iListbody);
+            var iReadOnlyListParam = new Parameter("innerList", null, typeof(IReadOnlyList<>), null, ValidationType.None, null);
+            var iReadOnlyListSignature = new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, new Parameter[] { iReadOnlyListParam });
+            var iReadOnlyListVariable = new ParameterReference(iReadOnlyListParam);
+            var iReadOnlyListbody = new MethodBodyStatement[]
+            {
+                new IfStatement(Snippets.NotEqual(iReadOnlyListVariable, Snippets.Null))
+                {
+                    new AssignValueStatement(_innerList, Snippets.Linq.ToList(iReadOnlyListVariable))
+                }
+            };
+
+            yield return new Method(iReadOnlyListSignature, iReadOnlyListbody);
+        }
+
+        protected override IEnumerable<CSharpType> BuildTypeArguments()
+        {
+            yield return _t.Type;
+        }
+
+        protected override IEnumerable<CSharpType> BuildImplements()
+        {
+            yield return new CSharpType(typeof(IList<>));
```

Comment:
since this interface should use the same argument of the generic parameter of this type itself, I suggest we change it this way:

```suggestion
            yield return new CSharpType(typeof(IList<>), _t.Type);
```

Reply:
I will try this, right now the writer automatically adds `T` but to your point if I wanted to call it TMyName then we would want that to flow through.